FingerBlox/app/src/main/java/org/fingerblox/fingerblox/ImageProcessing.java

```diff
@@ -131,10 +137,99 @@ else if(cn == 3)
             result.put(m.y-1, m.x  , color);
             result.put(m.y+1, m.x  , color);
         }
-        minutiaeField = filteredMinutiae;
+        MatOfKeyPoint keypoints = new MatOfKeyPoint();
+        keypoints.fromArray(minutiaeToKeyPoints(skeleton, filteredMinutiae));
+        keypointsField = keypoints;
+        DescriptorExtractor brief = DescriptorExtractor.create(DescriptorExtractor.ORB);
+        Mat descriptors = new Mat();
+        brief.compute(skeleton, keypoints, descriptors);
+        descriptorsField = descriptors;
+        return result;
+    }
+
+    private KeyPoint[] minutiaeToKeyPoints(Mat skeleton, HashSet<Minutiae> minutiae) {
+        KeyPoint[] result = new KeyPoint[minutiae.size()];
+        int index = 0;
+        float size = 1;
+        for (Minutiae m : minutiae) {
+            KeyPoint k;
+            float angle = -1;
+            float response = 1;
+            int octave = 1;
+            int class_id;
+            if (m.type == Minutiae.Type.RIDGEENDING) {
+                angle = getMinutiaeAngle(skeleton, m);
+                class_id = Minutiae.RIDGE_ENDING_LABEL;
+            } else {
+                class_id = Minutiae.BIFURCATION_LABEL;
+            }
+            result[index] = new KeyPoint(m.x, m.y, size, angle, response, octave, class_id);
+            index++;
+        }
         return result;
     }
 
+    private float getMinutiaeAngle(Mat skeleton, Minutiae m) {
+        Minutiae direction;
+        try {
+            direction = followRidge(skeleton, 5, m.x, m.y, m.x, m.y);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return -1;
+        }
+        double length = Math.sqrt(Math.pow(direction.x-m.x, 2) + Math.pow(direction.y-m.y, 2));
+        double cosine = (direction.y-m.y) / length;
+        float angle = (float)Math.acos(cosine);
+        if (direction.x - m.x < 0)
+            angle = -angle + 2 * (float)Math.PI;
+        return angle;
+    }
+
+    private Minutiae followRidge(Mat skeleton, int length, int currentX, int currentY, int previousX, int previousY) throws Exception {
+        if (length == 0)
+            return new Minutiae(currentX, currentY, Minutiae.Type.RIDGEENDING);
+        if (currentY >= skeleton.rows() - 1 || currentY == 0 || currentX >= skeleton.cols() - 1 || currentX == 0)
```

Comment:
assert crashes the application, right? I'm raising an exception since it's catched inside getMinutiaeAngle which then returns -1 (= angle is not applicable to this point)